File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    let [slow, fast] = [head, head];

    let count = 1;
    while(count <= n) {
        fast = fast.next;
        count++;
    }


    if (!fast) {
        return head.next;
    }

    while (fast.next){
        slow = slow.next;
        fast = fast.next;
    }

    console.log("slow", slow)
    console.log("fast", fast)

    slow.next = slow.next.next;

    return head;


    
};

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)

        count = 1
        curr = head
        while curr:
            curr = curr.next
            count += 1
       
        target = count - n

        count = 1
        curr = dummy
        while curr and count < target:
            curr = curr.next
            count += 1
       
        if curr.next:
            curr.next = curr.next.next
        else:
            curr.next = null
       
        return dummy.next
